File: sir_dev.R
#Gillespie SSA to run SIR model on static network (simulation of time-homogeneous CTMC on static network medium)
sir_homogeneous <- function(n_nodes, edge, root, beta, mu, t_end, info=TRUE, detail=FALSE){
  
  output <- vector(mode="list",length=t_end)
  
  #return vector of node states after each transition
  if(detail){
    delta_t <- 0.0
    detail_out <- vector(mode="list",length=5e3L)
    detail_i <- 1L
  }  

  x <- rep("s",n_nodes) #set node states to S 
  x[root] <- "i" #set state of root node to I
  
  m_I <- root #list of infected nodes
  N_I <- 1 #number of infected nodes
  N_R <- 0 #number of recovered nodes
  Mu <- mu #cumulative recovery rate
  tau <- rexp(n=1,rate=1) #draw tau Exp(1)
  
  #run through the time-steps:
  for(t in 1:t_end){
    
    #print simulation diagnostics
    if(info){
      print(paste0("current t: ",t,", number infected: ",N_I,", number recovered: ",N_R,", number susceptible: ",n_nodes-N_I-N_R))
    }
    
    #update list of possible S to I transitions
    m_SI <- NULL #S nodes in contact with I nodes
    for(k in 1:nrow(edge)){
      i <- edge[k,1]
      j <- edge[k,2]
      if(x[i] == "s" & x[j] == "i"){
        m_SI <- c(m_SI,i)
      } else if(x[i] == "i" & x[j] == "s"){
        m_SI <- c(m_SI,j)
      } else {
        next()
      }
    } #end for
    
    M_SI <- length(m_SI)
    Beta <- beta*M_SI #cumulative infection rate
    Lambda <- Mu + Beta #cumulative transition rate
    
    #check if a transition takes place
    if(tau >= Lambda){ #no transition
      tau = tau - Lambda
    } else { #at least one transition
      xi = 1.0 #remaining fraction of time-step
      while(tau < xi*Lambda){
        xi <- xi - tau/Lambda #fraction of time-step left after transition
        z <- runif(n=1,min=0,max=Lambda-.Machine$double.eps)
        if(z < Beta){ #S to I transition
          if(length(m_SI)==1){ #drawn m at random from m_SI
            m <- m_SI
          } else {
            m <- sample(x=m_SI,size=1) 
          }
          x[m] <- "i"
          m_I <- c(m_I,m)
          N_I <- N_I + 1
          Mu <- Mu + mu
        } else { #I to R transition
          if(length(m_I)==1){ #draw m at random from m_I
            m <- m_I
          } else {
            m <- sample(x=m_I,size=1) 
          }
          x[m] <- "r"
          m_I <- m_I[-which(m_I==m)] #remove m from m_I
          N_I <- N_I - 1
          N_R <- N_R + 1
          Mu <- Mu - mu
        } #end if
        
        #update list of S to I transitions and rates
        m_SI <- NULL #S nodes in contact with I nodes
        for(k in 1:nrow(edge)){
          i <- edge[k,1]
          j <- edge[k,2]
          if(x[i] == "s" & x[j] == "i"){
            m_SI <- c(m_SI,i)
          } else if(x[i] == "i" & x[j] == "s"){
            m_SI <- c(m_SI,j)
          } else {
            next()
          }
        } #end for
        
        #return detailed output
        if(detail){
          if(detail_i > length(detail_out)){ #expand list
            detail_out <- c(detail_out,vector(mode="list",length=100))
          }
          detail_out[[detail_i]]$x <- x
          detail_out[[detail_i]]$time <- t + (tau / Lambda)
          detail_i <- detail_i + 1L
        }
        
        M_SI <- length(m_SI)
        Beta <- beta*M_SI #cumulative infection rate
        Lambda <- Mu + Beta #cumulative transition rate
        
        tau = rexp(n=1,rate=1) #draw new tau
        
        #DEBUG
        if(n_nodes-N_I-N_R < 0){
          browser()
        }
        
      } #end while
    } #end if
    
    #write out desired quantities
    output[[t]]$x <- x
    output[[t]]$m_SI <- m_SI
    output[[t]]$m_I <- m_I
    output[[t]]$N_I <- N_I
    output[[t]]$N_R <- N_R
    
    if(N_I==0){
      break()
    }
    
  } #end for
  
  #return output
  if(detail){
    return(list(output=Filter(Negate(is.null),output),
                detail_out=Filter(Negate(is.null),detail_out)))
  } else {
    return(Filter(Negate(is.null),output))
  }
}


#generate the network medium the simulation will be run on
library(igraph)
n_nodes <- 200
erdos <- erdos.renyi.game(n=n_nodes,p=0.025,directed=FALSE,loops=FALSE)
erdos_edge <- as_edgelist(erdos)
erdos_edge <- erdos_edge[order(erdos_edge[,1]),]

beta <- 1.015/5 # R0 / infectious duration
mu <- 1/5 # 1 / infectious duration
set.seed(1)
sir_out <- sir_homogeneous(n_nodes=n_nodes,edge=erdos_edge,root=5,beta=beta,mu=mu,
                           t_end=50,info=TRUE,detail=TRUE)

sir_out_dat <- matrix(NA,nrow=length(sir_out$output),ncol=3)
sir_out_dat[,1] <- sapply(sir_out$output,function(x){n_nodes - x$N_I - x$N_R}) #S
sir_out_dat[,2] <- sapply(sir_out$output,function(x){x$N_I}) #I
sir_out_dat[,3] <- sapply(sir_out$output,function(x){x$N_R}) #R

get_state <- function(x){
  tab <- table(x)
  if(!"s" %in% names(tab)){
    tab[["s"]] <- 0
  }
  if(!"i" %in% names(tab)){
    tab[["i"]] <- 0
  }
  if(!"r" %in% names(tab)){
    tab[["r"]] <- 0
  }
  return(tab[c("s","i","r")])
}
sir_out_detail <- t(sapply(sir_out$detail_out,function(x){
  get_state(x$x)
}))
sir_out_detail_time <- sapply(sir_out$detail_out,function(x){x$time})

par(mfrow=c(1,2))
matplot(sir_out_dat,type="l",ylab="",main="Gillespie SSA on Static Network")
matplot(sir_out_detail_time,sir_out_detail,type="l",xlab="",ylab="",main="Gillespie SSA on Static Network (Detailed Output)")
par(mfrow=c(1,1))

File: sir_dev.cpp
#include <Rcpp.h>
// [[Rcpp::depends(RcppProgress)]]
#include <progress.hpp>

/*
 * node_tuple class stores information for two nodes (i,j):
 * node_tuple_list is a std::vector that will store the node tuples and function as an edgelist
 * that can be complemented with additional information (currently node indices and state)
 */
class node_tuple {
public:
  //constructor
  node_tuple() {} //empty constructor for initializing the continer vector
  node_tuple(int i, int j, char i_state, char j_state): i(i), j(j), i_state(i_state), j_state(j_state) {}
  //element assignment
  void set_i (int,char);
  void set_i_state (char);
  void set_j (int,char);
  void set_j_state (char);
  void set_all (int,int,char,char);
  //element accessors
  int get_i() {return i;}
  int get_j() {return j;}
  char get_i_state() {return i_state;}
  char get_j_state() {return j_state;}
private:
  int i; 
  int j; 
  char i_state; 
  char j_state;
};

//define element assignment methods
void node_tuple::set_i(int set_i, char set_i_state){
  i = set_i;
  i_state = set_i_state;
}
void node_tuple::set_j(int set_j, char set_j_state){
  j = set_j;
  j_state = set_j_state;
}
void node_tuple::set_all(int set_i, int set_j, char set_i_state, char set_j_state){
  i = set_i;
  i_state = set_i_state;
  j = set_j;
  j_state = set_j_state;
}
void node_tuple::set_i_state(char set_i_state){
  i_state = set_i_state;
}
void node_tuple::set_j_state(char set_j_state){
  j_state = set_j_state;
}


using namespace Rcpp;

//expose the node_tuple class to allow export to R
RCPP_EXPOSED_CLASS(node_tuple)
  RCPP_MODULE(mod_node_tuple){
    class_<node_tuple>("node_tuple")
    .constructor<int,int,char,char>("initialize node_tuple object")
    .method("set_i", &node_tuple::set_i,"sets node i attributes")
    .method("set_j", &node_tuple::set_j,"sets node j attributes")
    .method("set_all", &node_tuple::set_all,"sets node tuple attributes")
    .method("set_i_state", &node_tuple::set_i_state,"sets node i state")
    .method("set_j_state", &node_tuple::set_j_state,"sets node j state")
    .method("get_i", &node_tuple::get_i,"gets node i index")
    .method("get_j", &node_tuple::get_j,"gets node j index")
    .method("get_i_state", &node_tuple::get_i_state,"gets node i state")
    .method("get_j_state", &node_tuple::get_j_state,"gets node j state")
    ;
  }

//node_tuple_list functions as an edgelist (contactList in PLoS paper)
typedef std::vector<node_tuple> node_tuple_list;



/*
 * Gillespie algorithm to simulate SIR model on static network medium.
 * it will simulate a single trajectory of a time-homogeneous CTMC process.
 * THIS IS PURELY FOR DEVELOPMENT
 */

//grab number of unique nodes from edge list
int num_unique(NumericMatrix input){
  NumericVector elements;
  for(int i=0;i<input.nrow();i++){
    elements.push_back(input(i,0));
    elements.push_back(input(i,1));
  }
  std::sort(elements.begin(), elements.end());
  std::unique(elements.begin(), elements.end());
  int n_unique = 0;
  int i = 0;
  while(elements(i) != elements(i+1)){
    n_unique += 1;
    i++;
  }
  return(n_unique);
}

//given an edge list and a root infected, generate a node_tuple_list object
node_tuple_list init_contactList(NumericMatrix edge, int root){
  //allocate contactList
  int n_edge = edge.nrow();
  node_tuple_list contactList(n_edge);
  
  for(int i=0; i<n_edge; i++){
    
    //generate tuple i
    int i_index = edge(i,0);
    int j_index = edge(i,1);
    node_tuple tuple_i (i_index,j_index,'s','s');
    //check for root node
    if(i_index == root){
      tuple_i.set_i_state('i');
    }
    if(j_index == root){
      tuple_i.set_j_state('i');
    }
    contactList[i] = tuple_i;
    
  }
  
  return(contactList);
}

//given a contactList (a node_tuple_list object), generate another node_tuple_list that stores
//all of the tuples where a SI transition is possible
node_tuple_list generate_si_list(node_tuple_list contactList){
  
  node_tuple_list si_list;
  
  for(int i=0; i<contactList.size(); i++){
    
    //if neither infectious go to next iteration of loop;
    //dont want to store any combination of S or R people
    if(contactList[i].get_i_state() != 'i' && contactList[i].get_j_state() != 'i'){
      continue;
    } else {
      si_list.push_back(contactList[i]);
    }
    
  } //end loop
  
  return(si_list);
}

/*
 * note root uses 0 subsetting
 * beta: base infection rate
 * mu: base recovery rate
 */
void sir_homogeneous(NumericMatrix edge, int root, double beta, double mu, int t_end){

  //number of nodes
  int n_nodes = num_unique(edge);

  //initialize vector of node states
  CharacterVector node_states;
  for(int i=0;i<n_nodes;i++){
    node_states.push_back('s');
  }
  Rcout << node_states << std::endl;//DEBUGGING
  //set state of root node to i
  node_states(root) = 'i';

  //number of nodes in each state
  int n_inf = 1;
  int n_rec = 0;
  int n_sus = n_nodes - n_inf;

  //draw initial tau
  double tau = R::rexp(1.0);

  //initialize contactList
  node_tuple_list contactList = init_contactList(edge,root);

  //main simulation loop
  double time = 0.0; //time
  int iter = 0; //iterator
  double beta_cum; //cumulative infection rate
  double lambda_cum; //cumulative transition rate

  while(time <= t_end){
    
    //generate list of possible SI transitions
    node_tuple_list si_list = generate_si_list(contactList);
    
    int si_n = si_list.size(); //number of SI tuples
    beta_cum = si_n * beta; //cumulative infection rate
    lambda_cum = beta_cum + mu; //cumulative transition rate
    
    // Check if transition takes place during time-step:
    if(tau >= lambda_cum){ //no transition takes place
      tau -= lambda_cum;
    } else { //at least one transition took place
      
    }
    
    iter++; //update iterator
  }

}